### PR TITLE
enabling propagation of pdf uncertainties on moments and updating plo…

### PR DESCRIPTION
**Pull request by emanca**
…tting script

### CODE REVIEW

File: wremnants/syst_tools.py [hunk 1]
```diff
@@ -318,6 +320,16 @@ def add_pdf_hists(results, df, dataset, axes, cols, pdfs, base_name="nominal", a
         else:
             pdfHist = df.HistoBoost(pdfHistName, axes, [*cols, tensorName], tensor_axes=[pdf_ax], storage=hist.storage.Double())
             alphaSHist = df.HistoBoost(alphaSHistName, axes, [*cols, tensorASName], tensor_axes=[as_ax], storage=hist.storage.Double())
+
+            if propagateToHelicity:
+                pdfhelper = ROOT.wrem.makeHelicityMomentPdfTensor[npdf]()
```
Comment (kdlong) [inline wremnants/syst_tools.py:325]: Where is this defined? Is there a file missing in the commit?

Comment (emanca) [reply wremnants/syst_tools.py:325]: right... just added

File: wremnants/syst_tools.py [hunk 2]
```diff
@@ -2,13 +2,15 @@
...
 from utilities import boostHistHelpers as hh, common, logging
+import wremnants
 from wremnants import theory_tools
 from wremnants.datasets.datagroups import Datagroups
 from wremnants.helicity_utils import *
 import re
 import collections.abc
 
 logger = logging.child_logger(__name__)
+narf.clingutils.Declare('#include "theoryTools.h"')
```
Comment (cippy) [inline wremnants/syst_tools.py:13]: This should go in wremnants/\_\_init\_\_.py

File: utilities/differential.py [hunk 3]
```diff
@@ -56,13 +56,15 @@ def get_theoryAgnostic_axes():
 
     # Note that the helicity axis is defined elsewhere, and must not be added to the list of axes returned here
     axis_ptVgen = hist.axis.Variable(
-        [0., 5., 10., 15., 20., 25., 30., 35., 40., 45., 50.],
+        #[0., 5., 10., 15., 20., 25., 30., 35., 40., 45., 50.],
+        [0., 3., 6., 9.62315204,12.36966732,16.01207711,21.35210602,29.50001253,60.],
```
[the remark this answers is not in the document]
Comment (emanca) [reply utilities/differential.py:60]: yes agreed. @kdlong I guess the default binning should be the one for the theory corrections?

File: wremnants/syst_tools.py [hunk 4]
```diff
@@ -318,6 +320,16 @@ def add_pdf_hists(results, df, dataset, axes, cols, pdfs, base_name="nominal", a
         else:
             pdfHist = df.HistoBoost(pdfHistName, axes, [*cols, tensorName], tensor_axes=[pdf_ax], storage=hist.storage.Double())
             alphaSHist = df.HistoBoost(alphaSHistName, axes, [*cols, tensorASName], tensor_axes=[as_ax], storage=hist.storage.Double())
+
+            if propagateToHelicity:
```
Comment (davidwalter2) [inline wremnants/syst_tools.py:324]: What is the difference in this block with respect to the "if addhelicity" block in L313?

File: wremnants/include/theoryTools.h [hunk 5]
```diff
@@ -142,6 +142,56 @@ helicity_scale_tensor_t makeHelicityMomentScaleTensor(const CSVars &csvars, cons
...
+    const double sinPhiCS = csvars.sinphi;
+    const double cosPhiCS = csvars.cosphi;
+
+    const double sin2ThetaCS = 2. * sinThetaCS * cosThetaCS;
+    const double sin2PhiCS = 2. * sinPhiCS * cosPhiCS;
+    const double cos2ThetaCS = 1. - 2. * sinThetaCS * sinThetaCS;
+    const double cos2PhiCS = 1. - 2. * sinPhiCS * sinPhiCS;
+
+    // computing moments e.g. as used in arxiv:1708.00008 eq. 2.13
+    Eigen::TensorFixedSize<double, Eigen::Sizes<nhelicity, 1>> moments;
```
[the remark this answers is not in the document]
Comment (kdlong) [reply wremnants/include/theoryTools.h:170]: Yes let's recycle code here, possibly by inheriting from the other function

**Comment by emanca** [comment]
converting this open PR to draft to address comments and other WIP with @lucalavezzo and @federicovazzoler

File: utilities/differential.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cippy) [reply utilities/differential.py:60]: Perhaps we can discuss during the meeting, I imagine there is no rule to have it the same, but there might be an argument in having a consistent one. If this is agreed, what we could do is define this array of edges somewhere and import it for the default. In addition, we might add a couple of arguments to this function to accept a list of edges to override the default as needed (so the histomaker should get new options to set them, and internally propagate the choice here)

**Comment by kdlong** [comment]
I think the easiest way to recycle code would be to make this function work for both the moments helper and the application helper. It can be a separate function that is imported into both files, doesn't need to be a member of the class. 

https://github.com/WMass/WRemnants/blob/main/wremnants/include/theory_corrections.h#L16-L37

File: wremnants/include/theoryTools.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (emanca) [reply wremnants/include/theoryTools.h]: done

File: wremnants/syst_tools.py (hunk 4 above)
Comment (emanca) [reply wremnants/syst_tools.py:324]: removed as @cippy clarified that the usage of addHelicity was not what he intended it to be

**Comment by emanca** [comment]
After switching to the new scheme for angular moments and slightly different Collins-Soper definition the angular coefficients are now compatible with other publications (e.g. angular coefficients from Atlas). I slightly modified the definition of the helicity weights and verified closure, I.e. that all theory agnostic templates add up to the integrated distribution. I also implemented the first comments I received when I opened the PR. This PR passes all the checks and has no merge conflicts, therefore I’ll mark it as ready for review

File: narf
Comment (davidwalter2) [inline narf:1]: This updates narf to a newer version which could have things in development, @bendavid is this OK to update?

File: wremnants/helicity_utils.py [hunk 6]
```diff
@@ -23,25 +23,27 @@
...
         corrh = corrh[{'muRfact' : 1.j,}]
     if 'muFfact' in corrh.axes.name:
         corrh = corrh[{'muFfact' : 1.j,}]
-    missing = set(["y", "ptVgen", "chargeVgen", "helicity", "massVgen"]).difference(set(corrh.axes.name))
+    missing = set(["absYVgen", "ptVgen", "chargeVgen", "helicity", "massVgen"]).difference(set(corrh.axes.name))
     if missing != set():
         raise ValueError (f"Axes {missing} are not present in the coeff histogram")
     
-    corrh = corrh.project('massVgen','y','ptVgen','chargeVgen', 'helicity')
+    corrh = corrh.project('massVgen','absYVgen','ptVgen','chargeVgen', 'helicity')
```
Comment (davidwalter2) [inline wremnants/helicity_utils.py:45]: There was a previous PR #309 which had an unaddressed comment, could this be addressed?

**Review by davidwalter2 — COMMENTED** [review]
I think first the merge request in wremnants-data has to be merged. Would be good to also address the remaining comments.

**Comment by davidwalter2** [comment]
Hmm in the [PRValidation](https://cmsmwbot.web.cern.ch/WMassAnalysis/PRValidation/PR310/2023_10_19/impactsW.png) there are strong constraints on QCD scale uncertainties visible that were not there before, do you understand this?

**Comment by emanca** [comment]
> Hmm in the [PRValidation](https://cmsmwbot.web.cern.ch/WMassAnalysis/PRValidation/PR310/2023_10_19/impactsW.png) there are strong constraints on QCD scale uncertainties visible that were not there before, do you understand this?

I'll have a look but I suppose this comes from the fact that the helicity tensor changed slightly. I can try and check if it's correct, maybe I'll need @kdlong's help

File: wremnants/syst_tools.py (hunk 2 above)
Comment (emanca) [reply wremnants/syst_tools.py]: done

File: wremnants/helicity_utils.py (hunk 6 above)
Comment (emanca) [reply wremnants/helicity_utils.py]: done

**Comment by emanca** [comment]
> > Hmm in the [PRValidation](https://cmsmwbot.web.cern.ch/WMassAnalysis/PRValidation/PR310/2023_10_19/impactsW.png) there are strong constraints on QCD scale uncertainties visible that were not there before, do you understand this?
> 
> I'll have a look but I suppose this comes from the fact that the helicity tensor changed slightly. I can try and check if it's correct, maybe I'll need @kdlong's help

I've corrected it and it works locally. The CI is failing because it can't plot the impacts from the fit but this is not related to the code that I've changed with this PR....

**Comment by davidwalter2** [comment]
> > > Hmm in the [PRValidation](https://cmsmwbot.web.cern.ch/WMassAnalysis/PRValidation/PR310/2023_10_19/impactsW.png) there are strong constraints on QCD scale uncertainties visible that were not there before, do you understand this?
> > 
> > 
> > I'll have a look but I suppose this comes from the fact that the helicity tensor changed slightly. I can try and check if it's correct, maybe I'll need @kdlong's help
> 
> I've corrected it and it works locally. The CI is failing because it can't plot the impacts from the fit but this is not related to the code that I've changed with this PR....

No, the CI fails in the step before, when doing the fit. 
```status = 1, errstatus = 1, nllval = nan, nllvalfull = nan, edmval = -9.900000e+01, mineigval = -9.900000e+01, mineigvalinv = -9.900000e+01```
Then the impacts can not be plotted because the fit fails.

File: narf
Comment (bendavid) [reply narf:1]: No please don't update narf unless it's necessary.  The update will be included in my pr on xrootd handling.

Comment (emanca) [reply narf:1]: sorry this happened by mistake. Can you check that it is reverted now?

File: .gitmodules [hunk 7]
```diff
@@ -3,4 +3,4 @@
 	url = https://github.com/bendavid/narf.git
 [submodule "wremnants-data"]
 	path = wremnants-data
-	url = https://gitlab.cern.ch/cms-wmass/wremnants-data.git
+	url = https://gitlab.cern.ch/cms-wmass/wremnants-data.git
```
Comment (bendavid) [inline .gitmodules:6]: In principle this file shouldn't be touched by this PR anymore at all (I think this is just a trivial change of whitespace)

**Comment by bendavid** [comment]
Closing since the corresponding commits are included in https://github.com/WMass/WRemnants/pull/324

**Pull request closed without merging**